pkg/diagnostics/http_monitoring_path_matching.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package diagnostics
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+type pathMatching struct {
+	mux *http.ServeMux
+}
+
+func newPathMatching(paths []string, legacy bool) *pathMatching {
+	if paths == nil {
+		return &pathMatching{}
+	}
+
+	if len(paths) == 0 {
+		return &pathMatching{}
+	}
+
+	catchAllRegistered := false
+
+	mux := http.NewServeMux()
+	for _, pattern := range paths {
+		if pattern == "/" {
+			catchAllRegistered = true
+		}
+		mux.Handle(pattern, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			rw, ok := w.(*pathMatchingRW)
+			if !ok {
+				log.Errorf("Failed to cast to PathMatchingRW")
+				return
+			}
+			rw.MatchedPath = pattern
+		}))
+	}
+
+	// We can't register a catch-all handler if it was already registered
+	if !catchAllRegistered {
+		mux.Handle("/", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if !legacy {
+				rw, ok := w.(*pathMatchingRW)
+				if !ok {
+					log.Errorf("Failed to cast to PathMatchingRW")
+					return
+				}
+				rw.MatchedPath = unmatchedPathPlaceholder
+			}
+		}))
+	}
```

Comment:
Rather than have 2 `HandlerFunc` registrations, we can optionally append `"/"` if in non-legacy mode and remove dupes in the paths slice.


```go
if !legacy {
  paths = append(paths, "/")
  slices.Sort(paths)
  paths = slices.Compact(paths)
}
```


```suggestion
```

Reply:
I think we need the 2 handlers registration because we want to distinguish when the user explicitly sets the "/" path or not.

When the user sets the "/" path we match it and use it otherwise if the user does not set a "/" we fallback to the unmatched placeholder "_" so we need to have different handler funcs for those cases.

Reply:
done! we always need to register the catch all otherwise serveHTTP will throw an error when it doesn't match the path.

Reply:
done!